docs/envoy.md

```diff
@@ -0,0 +1,326 @@
+# Life of an OSM Request
+
+## A bit on Envoy
+
+OSM heavily uses envoy for proxying requests, applying mTLS, and exporting statistics. Envoy itself has 5 main
+configuration mechanisms for request manipulation:
+
+* ListenerConfigs
+* RouteConfigs
+* EndpointConfigs
+* ClusterConfigs
+* SecretConfigs
+
+### Listeners
+
+Envoy itself intercepts all requests by reconfiguring the IP tables on the Pod to reroute incoming and outbound traffic
+to the Sidecar. In OSM, incoming requests are intercepted and forwarded to port 15003, while outbound requests are
+forwarded to port 15001. Each port is configured with its own listener.
+
+In addition to the address and port to listen on, a Listener is configured with a set of `filter_chains` which dictate
+what to look for and what to do with a matching request, and a `listener_filter`, which tells Envoy what data it needs
+to extract from the request.
+
+Each Listener has a:
+
+
+Each listener will have a single `filter_chain` for every service. That is, the inbound listener will have 1 filter
+chain for each service that can reference its associated Pod, and the outbound filter will have 1 filter chain for every
+service the Pod can reach. A filter chain is comprised of 2 parts, a `filter_chain_match`, and a list of `filters`. This
+gives us the following construct:
+
+1. ListenerFilters
+2. FilterChains
+    1. Filter chain matchces
+    2. List of filters
+
+The main listener filter used in OSM is the "Original Destination" filter which tells the listener to pick up the
+address from a syscall (SO_ORIGINAL_DST) that gets the originally requested IP address from the socket (prior to the
+IP table redirect). Otherwise, the IP used in the filter chain match would be the the redirected IP referencing the
+sidecar itself.
+
+#### Example flow
+
+The flow of a request from `pod-1` backed by `service-a` to `service-b` is as follows:
+
+The pod performs a DNS lookup for `service-b`, which routes to the in-cluster DNS service (typically CoreDNS or KubeDNS)
+which will return the *service* ClusterIP address (we are currently ignoring the case for headless services, or
+ExternalName). This DNS request is permitted, and not redirected via the IP Tables redirects.
+
+Next, the pod will attempt to make a request to that IP address, on that port. The IP tables are setup to redirect the
+to port 15001, which the `outbound-listener` is listening on. The Listener is configured to extract the SO_ORIGINAL_DST,
+and passes that through the set of filter chains, looking for the most precise match.
+
+On a filter match, envoy will now apply that filter chain's full set of filters, 1 by 1, in order. The Filter we are
+interested in, is called the `HttpConnectionManager`, which uses rDS to pass apply a RouteConfiguration, which OSM calls
```

Comment:
Do you think that mentioning that `HttpConnectionManager` is an http (l7) proxy filter might make this more easier to read for someone not familiar with Envoy docs?